metricbeat/scripts/mage/docsCollector.go

```diff
@@ -0,0 +1,488 @@
+		}
+		cfgFile = cfgBack
+	} else {
+		return "", fmt.Errorf("config file in %s could not be found", modulePath)
+	}
+
+	raw, err := ioutil.ReadFile(cfgFile)
+	return string(raw), err
+
+}
+
+//gather all the metricsets for a given module
+func gatherMetricsets(modulePath string, moduleName string) ([]metricsetData, error) {
+	metricsetList, err := filepath.Glob(filepath.Join(modulePath, "/*"))
+	if err != nil {
+		return nil, err
+	}
+	var metricsets []metricsetData
+	for _, metricset := range metricsetList {
+		isMetricset, err := testIfDocsInDir(metricset)
+		if err != nil {
+			return nil, err
+		}
+		if !isMetricset {
+			continue
+		}
+		metricsetName := filepath.Base(metricset)
+		release, err := getReleaseState(filepath.Join(metricset, "_meta/fields.yml"))
+		if err != nil {
+			return nil, err
+		}
+
+		//generate the asciidoc link used in the module docs, since we need this in a few places
+		link := fmt.Sprintf("<<metricbeat-metricset-%s-%s,%s>>", moduleName, metricsetName, metricsetName)
+
+		//test to see if the metricset has a data.json
+		hasData := false
+		_, err = os.Stat(filepath.Join(metricset, "_meta/data.json"))
+		if err == nil {
+			hasData = true
+		}
+
+		ms := metricsetData{
+			Path:       metricset,
+			Title:      metricsetName,
+			Release:    release,
+			Link:       link,
+			DataExists: hasData,
+		}
+
+		metricsets = append(metricsets, ms)
+
+	} // end of metricset loop
+
+	return metricsets, nil
+}
+
+//Gather all the data we need to construct the docs that end up in metricbeat/docs
+func gatherData(modules []string) (map[string]moduleData, error) {
+	moduleMap := make(map[string]moduleData)
+	//iterate over all the modules, checking to make sure we have an asciidoc file
+	for _, module := range modules {
+
+		isModule, err := testIfDocsInDir(module)
+		if err != nil {
+			return moduleMap, err
+		}
+		if !isModule {
+			continue
+		}
+		moduleName := filepath.Base(module)
+
+		err = createDocsPath(moduleName)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		fieldsm, err := loadModuleFields(filepath.Join(module, "_meta/fields.yml"))
+		if err != nil {
+			return moduleMap, err
+		}
+
+		cfgPath, err := getConfigfile(module)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		metricsets, err := gatherMetricsets(module, moduleName)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		//dump the contents of the module asciidoc
+		moduleDoc, err := ioutil.ReadFile(filepath.Join(module, "_meta/docs.asciidoc"))
+		if err != nil {
+			return moduleMap, err
+		}
+
+		fieldsm.Path = module
+		fieldsm.CfgFile = cfgPath
+		fieldsm.Metricsets = metricsets
+		fieldsm.Asciidoc = string(moduleDoc)
+		fieldsm.Dashboards = hasDashboards(module)
+
+		moduleMap[moduleName] = fieldsm
+
+	} // end of modules loop
+
+	return moduleMap, nil
+}
+
+// write the module-level docs
+func writeModuleDocs(modules map[string]moduleData) error {
+	for moduleName, mod := range modules {
+
+		moduleFile := generatedNote
```

Comment:
> I chose this moduleFile = moduleFile thing over a giant Sprintf() as I thought it cuts down on oftentimes illegible printf directives, and makes it clear to the reader what a given line is actually doing. Is this the way we want to do this? I'm not hard-set on it.

I think I'd use a [bytes buffer](https://golang.org/pkg/bytes/#Buffer) here.
```
var moduleFile bytes.Buffer
moduleFile.WriteString(fmt.Sprintf("[[metricbeat-module-%s]]\n", moduleName))
...
err := ioutil.WriteFile(filename, moduleFile.Bytes(), 744)
```

Another option is to move this logic to a template, but this would be a bigger change.

And the same in other methods where we concatenate formatted strings.